starfish/spots/_detector/combine_adjacent_features.py

```diff
@@ -14,120 +18,358 @@ class ConnectedComponentDecodingResult(NamedTuple):
+            target name (str) and feature radius.
+        int :
+            1 if spot passes size filters, zero otherwise.
+
+        """
+        # because of the above skimage issue, we need to support both 2d and 3d properties
+        if len(spot_property.centroid) == 3:
+            spot_attrs = {
+                'z': int(spot_property.centroid[0]),
+                'y': int(spot_property.centroid[1]),
+                'x': int(spot_property.centroid[2])
+            }
+        else:  # data is 2d
+            spot_attrs = {
+                'z': 0,
+                'y': int(spot_property.centroid[0]),
+                'x': int(spot_property.centroid[1])
+            }
+
+        # we're back to 3d or fake-3d here
+        target_index = decoded_image[spot_attrs['z'], spot_attrs['y'], spot_attrs['x']]
+        spot_attrs[Features.TARGET] = target_map.target_as_str(target_index)
+        spot_attrs[Features.SPOT_RADIUS] = spot_property.equivalent_diameter / 2
+
+        # filter intensities for which radius is too small
+        passes_area_filter = 1 if min_area <= spot_property.area < max_area else 0
+        return spot_attrs, passes_area_filter
+
+    def _create_spot_attributes(
+            self,
+            region_properties: List[_RegionProperties],
+            decoded_image: np.ndarray,
+            target_map: TargetsMap,
+            n_processes: Optional[int]=None
+    ) -> Tuple[SpotAttributes, np.ndarray]:
+        """
+
+        Parameters
+        ----------
+        region_properties : List[_RegionProperties]
+            Properties of the each connected component. Output of skimage.measure.regionprops
+        decoded_image : np.ndarray
+            Image whose pixels correspond to the targets that the given position in the ImageStack
+            decodes to.
+        target_map : TargetsMap
+            Unique mapping between string target names and int target IDs.
+        n_processes : Optional[int]=None
+            number of processes to devote to measuring spot properties. If None, defaults to the
+            result of os.nproc()
+
+        Returns
+        -------
+        pd.DataFrame :
+            DataFrame containing x, y, z, radius, and target name for each connected component
+            feature.
+        np.ndarray[bool] :
+            An array with length equal to the number of features. If zero, indicates that a feature
+            has failed area filters.
+        """
+        pool = Pool(n_processes)
+        mapfunc = pool.map
+        applyfunc = partial(
+            self._single_spot_attributes,
+            decoded_image=decoded_image,
+            target_map=target_map,
+            min_area=self._min_area,
+            max_area=self._max_area
+        )
+
+        iterable = tqdm(region_properties)
+        results = mapfunc(applyfunc, iterable)
+        spot_attrs, passes_area_filter = zip(*results)
```

Comment:
let's make sure that when we filter spot attribute tables, we're filtering on the same quantities across the different decoders. right now, the max_channel_per_hyb_round decoder filters on area and intensity. this decoder filters on area only. ALL decoders should filter on area, intensity, and quality/distance. this will allow for better unification of our API, pipeline, etc. Perhaps leave a more craftfully worded version of his as an Issue?

Reply:
Updated #499